### ISSUE TITLE
Upgrade Testing framework

### ISSUE THREAD

**Issue by samschurter** [materializecss/materialize#48]
In working on Dropdown I added a bunch of tests and ran into some issues. While investigating those issues I noticed that our testing framework is currently something like this, to the best of my knowledge. I'm pretty new to Grunt and front-end testing in general, so let me know if this isn't how it works.

1. grunt-contrib-jasmine (current version 2.20, Materialize version 1.2.0) is the Grunt plugin that uses PhantomJS to run Jasmine tests.
2. PhantomJS (current version 2.1, no longer developed) is the headless webkit simulating the browser for Jasmine to run in.
3. Jasmine (current version 3.6, materialize version 2.6) is the test runner and supplies most of the matchers.
4. Jasmine-jquery (current version 2.1.1 VERY abandoned) provides additional matchers for jQuery and the fixture loaders.

The only 2 components that aren't vastly out of date are because they haven't been developed in years.

It looks like the hardest thing to upgrade will be jasmine-jquery, so maybe a good plan would be to replace one component at a time and see what happens? Maybe something like 
- grunt-contrib-jasmine 1.2.0 -> 2.20 (will mean changing from PhantomJS to Chrome Headless)
- jasmine 2.6 -> 3.6 (I don't see any huge breaking changes in the v2 to v3 upgrade guide, so hopefully easy)
- jasmine-jquery -> ? I'm lost on this one. I'm not sure what options there even are. @tomwjerry mentioned using a server for the fixtures or directly making them using DOM creation in javascript.

As I added tests I started thinking maybe any upgrades should be done first. Upgrading could end up being a pretty big change, especially if we replace jasmine-jquery, and I feel like it would be better to do while we have about 60 test cases rather than hundreds. What do the rest of you think? Worth upgrading now before we write more tests, or wait until something else forces an upgrade?

**Comment by Smankusors** [comment]
hmm... if we do upgrade to the chrome headless, it's included when doing `npm install`, or we need to install it manually?

**Comment by samschurter** [comment]
Good question, I for one definitely want it to be as seamless as possible. It uses headless chrome via the [puppeteer project](https://github.com/puppeteer/puppeteer), which downloads a version of chromium guaranteed to work with puppeteer. I branched an upgraded just grunt-contrib-jasmine and it installed chrome and started up just fine, but the tests all fail because jasmine-jquery can't load the fixtures due to a CORS issue.

**Comment by samschurter** [comment]
I added a task to the grunt file to start a server for the fixtures and it's working.

**Comment by samschurter** [comment]
Upgrading to Jasmine 3.6 randomly breaks some tests (FAB, and Toast I have run into so far). 3.6 defaults to running tests in random order and turning random order off makes all the tests pass. So there's some tests in there that are dependent on order and need to be fixed, otherwise the upgrade from grunt-contrib-jasmine 1.2 -> 2.2 and jasmine 2.6 -> 3.6 is pretty easy.

Next I'm looking at replacing the jasmine-jquery functionality and probably stripping out all jQuery. Materialize isn't supposed to depend on jQuery anymore but there have been reports of things only working with jQuery. As long as the test suite is loading jQuery I don't see a way to make sure the components work without it.

**Comment by DanielRuf** [comment]
Can you provide a (draft) PR so we can discuss the best route / solutions?

**Comment by tomwjerry** [comment]
I have ported all the tests from jQuery (quite a bit of change) to vanilla js based on @samschurter code. I don't want to steal his work but I based my work on his repo since some tests has been added and he did some changes to use a local server for testing. sam should probably answer on this but in order to make things going I can provide a PR after waiting a little. Since there are many changes it is going to take some time to review.
